tests/unit/db/database_tests.py

```diff
@@ -0,0 +1,480 @@
+            reduce=False)
+        self.assertIsInstance(rslt, Result)
+        self.assertEqual(rslt[:1], rslt['julia099'])
+
+    def test_retrieve_raw_view_results(self):
+        """
+        Test retrieving raw output from a design document view
+        """
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        data = {'_id': '_design/ddoc01','views': {'view01': {"map": map_func}}}
+        self.db.create_document(data)
+        self.populate_db_with_documents()
+
+        raw_rslt = self.db.get_view_raw_result('_design/ddoc01', 'view01')
+        self.assertIsInstance(raw_rslt, dict)
+        self.assertEqual(len(raw_rslt.get('rows')), 100)
+
+    def test_all_docs(self):
+        """
+        Test the all_docs functionality
+        """
+        self.populate_db_with_documents()
+        data = self.db.all_docs(
+            limit=3,
+            keys=['julia006', 'julia024', 'julia045', 'julia099']
+        )
+        self.assertEqual(len(data.get('rows')), 3)
+        self.assertEqual(data['rows'][0]['key'], 'julia006')
+        self.assertEqual(data['rows'][1]['key'], 'julia024')
+        self.assertEqual(data['rows'][2]['key'], 'julia045')
+
+    def test_custom_result_context_manager(self):
+        """
+        Test using the database custom result context manager
+        """
+        self.populate_db_with_documents()
+        with self.db.custom_result(startkey='julia010', endkey='julia012') as rslt:
+            self.assertIsInstance(rslt, Result)
+            keys_returned = []
+            for i in rslt:
+                keys_returned.append(i['key'])
+            expected_keys = ['julia010', 'julia011', 'julia012']
+            self.assertTrue(all(x in keys_returned for x in expected_keys))
+
+    def test_keys(self):
+        """
+        Test retrieving the document keys from the database
+        """
+        self.assertEqual(self.db.keys(), [])
+        self.populate_db_with_documents(3)
+        self.assertEqual(
+            self.db.keys(remote=True),
+            ['julia000', 'julia001', 'julia002']
+            )
+
+    def test_get_non_existing_doc_via_getitem(self):
+        """
+        Test __getitem__ when retrieving a non-existing document
+        """
+        try:
+            doc = self.db['no_such_doc']
+            self.fail('Above statement should raise a KeyError')
+        except KeyError:
+            pass
+
+    def test_get_db_via_getitem(self):
+        """
+        Test __getitem__ when retrieving a document
+        """
+
+        # Add a design document
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        expected_ddoc = self.db.get_design_document('_design/ddoc01')
+        expected_ddoc.add_view('view01', map_func)
+        expected_ddoc.save()
+
+        # Add three standard documents
+        self.populate_db_with_documents(3)
+
+        # Test __get_item__ for standard document
+        doc = self.db['julia001']
+        self.assertIsInstance(doc, Document)
```

Comment:
This set of assertions is repeated in a lot of places. Does it make sense to build a function in order to be more DRY? Or maybe being explicit is better.